### PR TITLE
add nix-shell boilerplate

### PR DESCRIPTION
**Pull request by Xe**
This enables users of nix-shell and lorri to automagically have the correct development environment by simply changing directory into a checkout of this repo. For more information on this see the following links:

- https://christine.website/blog/how-i-start-nix-2020-03-08
- https://direnv.net/
- https://www.tweag.io/blog/2019-03-28-introducing-lorri/

Signed-off-by: Christine Dodrill <xe@tailscale.com>

### CODE REVIEW

**Review by crawshaw — COMMENTED** [review]
I think this would be the first env-specific code added to the open source repository. I'm a little surprised this is necessary, as we have worked hard to keep this repository as "vanilla Go" as possible. Do all Go projects need this for Nix? Is there no standard way to work in a typical Go repository in Nix?

File: .envrc [hunk 1]
```diff
@@ -0,0 +1 @@
+eval "$(lorri direnv)"
```
Comment (danderson) [inline .envrc:1]: I would not submit .envrc. That leaves it to each individual how they want their environment to work (e.g. I use nix-direnv and `use nix` instead of lorri, for reasons). That's also what I did in the corp repo, shell.nix is there if you want it but it's up to you how you enable it in your environment.

File: shell.nix [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{ pkgs ? import <nixpkgs> {} }:
+
+pkgs.mkShell {
+  buildInputs = with pkgs; [
+    go goimports gopls
+
+    # keep this line if you use bash
+    pkgs.bashInteractive
```
Comment (danderson) [inline shell.nix:8]: Why? I've never included shells in my shell.nix, and it just inherits from its environment. What do we gain from this?

**Review by danderson — APPROVED** [review]
IMHO, shell.nix (without .envrc) is harmless to those who don't care, but has a specific value to third parties: it tells them what our tooling stack is, without having to write a bitrotting markdown file explaining it. As such, I'm happy to add it (although most of the value in our case comes from the corp repo, because its dependency stack is much weirder)

File: shell.nix [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{ pkgs ? import <nixpkgs> {} }:
+
+pkgs.mkShell {
+  buildInputs = with pkgs; [
+    go goimports gopls
```
Comment (danderson) [inline shell.nix:5]: Missing git, which is invoked explicitly in some build paths to embed release info.

**Review by danderson — APPROVED** [review]

File: shell.nix (hunk 2 above)
Comment (Xe) [reply shell.nix:8]: This was something lorri added for me. I'll remove it.

**Comment by Xe** [comment]
@crawshaw the Nix ethos is all about being very explicit on what depends on what, as such there is no real "standard" go setup. A nice side effect of this is that it makes you very conscious about what dependencies are being pulled in everywhere (and not just in development environments).

**Comment by crawshaw** [comment]
To phrase my question more generally: if we receive PRs adding environmental setup scripts for WSL, Debian, homebrew, etc, do we accept them? This is the first env-specific code in this repository, what should our policy on this be?

**Comment by crawshaw** [comment]
If this isn't blocking you, I wouldn't mind if we could wait on submitting this so I can hear env policy opinions from @danderson and @bradfitz.

We do already have a Dockerfile, which I suppose is another example of an env-specific script.

**Comment by danderson** [comment]
I think debian and homebrew are separate questions, because those are package repositories, so that would be the equivalent of our debian and rpm packaging logic. For those, the question boils down to: who owns maintaining those? If it's us, it should go here, unless the package system has different conventions (for example, I maintain the Arch and Nix packages for tailscale, and they're hosted in the distro-canonical places for those packages).

This shell.nix is a development environment setup. In that sense, it's equivalent to one possible Dockerfile we could ship (which isn't the one we currently ship). As it happens it's portable across any linux and macOS system that has Nix installed, and can be used as a shortcut for "what tools do you need to build this repository?" for those folks. For everyone else, it's a no-op that can be ignored.

The only other such dev environment I could think of supporting at some point is either something Windows-specific, or WSL-specific. I don't know enough about those universes to know if that would ever make sense though.

In conclusion: I'm willing to set the precedent that non-intrusive environment setup things that we (broadly, tailscale contributors) use are okay to live in this repository. Things we don't personally use would come down to a judgement call about the benefit to the community (i.e. if it's something to do with homebrew, that probably benefits enough people; if it's "danderson's custom shell script of doom", probably not).

**Comment by bradfitz** [comment]
I'm fine with this as long as a comment is added at the top with some context for nix noobs saying what it is (it's for Nix!), how you'd use it, a link to some background reading docs, etc.

**Comment by Xe** [comment]
@crawshaw @danderson @bradfitz I attempted to comment things out in https://github.com/tailscale/tailscale/pull/1028/commits/76abe6b6e363e99b161fb562b883f6266093ba51, how is that?

**Comment by bradfitz** [comment]
LGTM. Squash & merge away.
